src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,674 @@
+
+        read_request = await self.default_controller.ReadAttribute(self.dut_node_id, [(0, Clusters.Objects.Descriptor)])
+
+        asserts.assert_equal({0}, read_request.keys(), "Endpoint 0 not in output")
+        asserts.assert_equal({Clusters.Descriptor}, read_request[0].keys(), "Descriptor cluster not in output")
+        attribute_ids = [a.attribute_id for a in read_request[0][Clusters.Descriptor].keys() if a != Clusters.Attribute.DataVersion]
+        returned_attributes = read_request[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.AttributeList]
+        asserts.assert_equal(sorted(attribute_ids), sorted(returned_attributes), "Expected attribute list doesn't match")
+
+        # Step 3
+
+        # TH sends the Read Request Message to the DUT to read an attribute from a cluster at all Endpoints
+        # AttributePath = [[Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the Endpoints to the DUT.
+
+        self.print_step(3, "Send Request Message to read one attribute on a given cluster at all endpoints")
+        all_attributes = await self.default_controller.ReadAttribute(self.dut_node_id, [Clusters.Objects.Descriptor])
+        for endpoint in all_attributes:
+            asserts.assert_in(Clusters.Objects.Descriptor,
+                              all_attributes[endpoint], f"Descriptor attribute not found in endpoint: {endpoint}")
+            attribute_ids = [a.attribute_id for a in all_attributes[endpoint]
+                             [Clusters.Descriptor].keys() if a != Clusters.Attribute.DataVersion]
+            returned_attributes = all_attributes[endpoint][Clusters.Descriptor][Clusters.Descriptor.Attributes.AttributeList]
+            asserts.assert_equal(sorted(attribute_ids), sorted(returned_attributes), "Expected attribute list doesn't match")
+
+        # Step 4
+
+        # TH sends the Read Request Message to the DUT to read a global attribute from all clusters at that Endpoint
+        # AttributePath = [[Endpoint = Specific Endpoint, Attribute = Specific Global Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the clusters to the DUT.
+        self.print_step(4, "Send Request Message to read one global attribute from all clusters at that endpoint")
+        attribute_path_1 = AttributePath(
+            EndpointId=0,
+            ClusterId=None,
+            AttributeId=global_attribute_ids.GlobalAttributeIds.ATTRIBUTE_LIST_ID)
+
+        read_request = await self.default_controller.ReadAttribute(
+            self.dut_node_id,
+            [attribute_path_1]
+        )
+
+        for endpoint in read_request:
+            asserts.assert_in(Clusters.Objects.Descriptor, read_request[endpoint].keys(), "Descriptor cluster not in output")
```

Comment:
This is verifying that the descriptor cluster attribute list appears in all endpoints, but the read was for all clusters at endpoint 0. So read_request will have only 1 endpoint, and you're only verifying one cluster. This needs to verify against the server list as noted in the comment below.

Reply:
Reworked to only use endpoint 0